graph-builder/src/config/file.rs

```diff
@@ -0,0 +1,99 @@
+//! TOML file configuration options.
+
+use failure::{Fallible, ResultExt};
+use std::collections::{HashMap, HashSet};
+use std::io::Read;
+use std::{fs, io, path};
+
+/// TOML configuration, top-level.
+#[derive(Debug, Deserialize)]
+pub struct FileOptions {
+    /// Verbosity level.
+    pub verbosity: Option<u8>,
+
+    /// Upstream options.
+    pub upstream: Option<UpstreamOptions>,
+
+    /// Web frontend options.
+    pub service: Option<ServiceOptions>,
+
+    /// Status service options.
+    pub status: Option<StatusOptions>,
+
+    /// Plugins ordering and options.
+    pub plugins: Option<Vec<HashMap<String, String>>>,
+}
+
+impl FileOptions {
+    pub fn read_filepath<P: AsRef<path::Path>>(cfg_path: P) -> Fallible<Self> {
+        let cfg_file = fs::File::open(&cfg_path).context(format!(
+            "failed to open config path {:?}",
+            cfg_path.as_ref()
+        ))?;
+        let mut bufrd = io::BufReader::new(cfg_file);
+
+        let mut content = vec![];
+        bufrd.read_to_end(&mut content)?;
+        let cfg = toml::from_slice(&content).context(format!(
+            "failed to read config file {}",
+            cfg_path.as_ref().display()
+        ))?;
```

Comment:
Not directly because the expected error type is a `failure::Error` while this is a `failure::Context`. I could chain some error-mapping to massage types, but I think it's more straightforward to read in its current form.

Reply:
So the `?` handles that conversion automatically?

Reply:
Eventually everything converts to a `Result` and `std::error::Error` trait. There a few layers to go through in `failure`, though. I don't know the exact path that `?` goes through, but it is pretty effective at doing those conversions automatically, yes.